tests/diff-folded.rs

```diff
@@ -0,0 +1,159 @@
+#[macro_use]
+extern crate pretty_assertions;
+
+extern crate inferno;
+
+use inferno::diff_folded::{self, Options};
+use log::Level;
+use std::fs::{self, File};
+use std::io::{self, BufRead, BufReader, Cursor};
+
+fn test_diff_folded(
+    infile1: &str,
+    infile2: &str,
+    expected_result_file: &str,
+    options: Options,
+) -> io::Result<()> {
+    let metadata = match fs::metadata(expected_result_file) {
+        Ok(m) => m,
+        Err(e) => {
+            if e.kind() == io::ErrorKind::NotFound {
+                // be nice to the dev and make the file
+                let mut f = File::create(expected_result_file).unwrap();
+                diff_folded::from_files(&options, &infile1, &infile2, &mut f)?;
+                fs::metadata(expected_result_file).unwrap()
+            } else {
+                return Err(e.into());
+            }
+        }
+    };
+
+    let expected_len = metadata.len() as usize;
+    let mut result = Cursor::new(Vec::with_capacity(expected_len));
+    let return_value = diff_folded::from_files(&options, infile1, infile2, &mut result)?;
+    let expected = BufReader::new(File::open(expected_result_file).unwrap());
+    // write out the expected result to /tmp for easy restoration
+    result.set_position(0);
+    let rand: u64 = rand::random();
+    let tm = std::env::temp_dir().join(format!("test-{}.svg", rand));
+    if fs::write(&tm, result.get_ref()).is_ok() {
+        eprintln!("test output in {}", tm.display());
+    }
+    // and then compare
+    result.set_position(0);
+    compare_results(result, expected, expected_result_file);
+    Ok(return_value)
+}
+
+fn compare_results<R, E>(result: R, expected: E, expected_file: &str)
+where
+    R: BufRead,
+    E: BufRead,
+{
+    let result_lines: Result<Vec<String>, _> = result.lines().collect();
+    let mut result_lines = result_lines.unwrap();
+    let expected_lines: Result<Vec<String>, _> = expected.lines().collect();
+    let mut expected_lines = expected_lines.unwrap();
+
+    assert_eq!(
+        result_lines.len(),
+        expected_lines.len(),
+        "\nresult has {} lines, expected {} lines",
+        result_lines.len(),
+        expected_lines.len()
+    );
+
+    result_lines.sort_unstable();
+    expected_lines.sort_unstable();
+
+    for (line_num, (result_line, expected_line)) in result_lines
+        .into_iter()
+        .zip(expected_lines.into_iter())
+        .enumerate()
+    {
+        assert_eq!(
+            result_line, expected_line,
+            "\n{}:{}",
+            expected_file, line_num
+        );
+    }
+}
```

Comment:
This too I sort of wish we could extract out and share. Not sure how easy that'll be though..

Reply:
This one is quite a bit different than the other ones though because we sort the outputs before comparing. And because we have to read in all of the lines before we sort, we can compare the lengths of the outputs at the beginning.